Fixes #60989: add metrics for mark cache evicted bytes, marks and files

shiv4289 opened the pull request:
**Resolves** #60989

This PR adds detailed visibility into the behavior of the Mark Cache by introducing new `ProfileEvents` for tracking evictions. Specifically, it tracks:

- `MarkCacheEvictedBytes`  
- `MarkCacheEvictedMarks`  
- `MarkCacheEvictedFiles`

### Motivation

Previously, eviction tracking in the cache relied on a single callback invocation of `onRemoveOverflowWeightLoss()`:

```cpp
/// Override this method if you want to track how much weight was lost in removeOverflow method.
virtual void onRemoveOverflowWeightLoss(size_t /*weight_loss*/) {}
```
This callback only provided the total weight evicted (in bytes), aggregated in the cache policy implementation (`LRUCachePolicy.h`  or `SLRUCachePolicy.h`) and relayed back to the cache implementation class (`MarckCache.h`) after all evictions. It lacked contextual information about individual entries being removed.

### Changes Introduced
This patch replaces the `onRemoveOverflowWeightLoss()` callback with a more expressive method: `onEntryRemoval()`. The new method provides both the evicted entry and its weight at the time of removal:

```cpp
/// Called when an entry is evicted from the cache.
/// Override this to track individual entry removals.
virtual void onEntryRemoval(size_t weight_loss, const MappedPtr & entry) {}
```
This allows for more fine-grained eviction metrics currently implemented in `MarkCache.h.` 

The cache policy ( `LRUCachePolicy.h `or `SLRUCachePolicy.h`) is responsible for computing the weight of the entry. By passing this value to the callback, we avoid redundant computation within the cache implementation classes.

### Affected Components
All cache implementations have been updated to override the new `onEntryRemoval` callback:

- MarkCache  (`MarkCache.h`)
- PageCache  (`PageCache.h`, `PageCache.cpp`)
- UncompressedCache  (`UncompressedCache.h`)
- VectorSimilarityIndexCache  (`VectorSimilarityIndexCache.h`)
- IcebergMetadataFilesCache ( `IcebergMetadataFilesCache.h`)

This groundwork enables future improvements in observability and diagnostics for all these caches.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added metrics `MarkCacheEvictedBytes`, `MarkCacheEvictedMarks`, `MarkCacheEvictedFiles` for tracking evictions from the mark cache. (issue #60989)

### Documentation entry for user-facing changes

The [documentation page](https://clickhouse.com/docs/operations/system-tables/events) is done in a way that does not require update.

shiv4289 committed:
Added Profile Event for MarkCacheEvictedBytes

shiv4289 committed:
Added ProfileEvent for MarkCacheEvictedFiles

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=80799) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=80799) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=80799) it.</sub>

shiv4289 committed:
Added Profile Event for MarkCacheEvictedMarks

kssenii commented:
> While I am looking for a neat way to write a test for all these metrices in a short-running test

As you'll need to check eviction metrics, then `stateless`  tests (`tests/functional/queries/0_stateless/`) is not an option, because it would be difficult to guarantee eviction there. Therefore the remaining option is `integration` tests (`tests/integration/`), where you can configure mark cache to be very small, therefore it would be easy to guarantee eviction and see the metrics you added. E.g. create a table, fill it, read from it with a specific query id and you will be able to check profile events in `system.query_log` in `ProfileEvents` column. Are you on linux or macOS? Running integration tests on macOS is tricky, but possible, so if you are using mac, then tell and I will elaborate on what to do. Otherwise just see `tests/integration/README.md`.
Also there is one more option - unit test (see files names `gtest_*`). But it is a less preferable option.

shiv4289 committed:
fix CI - cosmetic changes to match style

shiv4289 committed:
added integration tests for all mark cache eviction metrics

kssenii commented on src/Common/SLRUCachePolicy.h:
```diff
@@ -295,6 +296,7 @@ class SLRUCachePolicy : public ICachePolicy<Key, Mapped, HashFunction, WeightFun
             else
             {
                 current_weight_lost += cell.size;
+                evicted_values.push_back(cell.value); // Store the evicted value
```
Let's better immediately calculate `total_evicted_marks_num` and `total_evicted_files_num` instead of collecting evicted objects? And rename `EvictionDetails` to `EvictionStats`

shiv4289 replied:
Hi @kssenii , thank you for the suggestion! I had a question regarding the calculation of `total_evicted_marks_num` here.

When we [load MarkCache](https://github.com/ClickHouse/ClickHouse/blob/759349e6a2fb0f15506c5ff9abc71ac149042db1/src/Storages/MergeTree/MergeTreeMarksLoader.cpp#L212), each [cell](https://github.com/ClickHouse/ClickHouse/blob/759349e6a2fb0f15506c5ff9abc71ac149042db1/src/Common/SLRUCachePolicy.h#L225) in the SLRU cache stores a pointer to [MarksInCompressedFile](https://github.com/ClickHouse/ClickHouse/blob/759349e6a2fb0f15506c5ff9abc71ac149042db1/src/Formats/MarkInCompressedFile.h#L42).

Since `SLRUCachePolicy.h` is decoupled from `MarksInCompressedFile` [via templating](https://github.com/ClickHouse/ClickHouse/blob/759349e6a2fb0f15506c5ff9abc71ac149042db1/src/Common/SLRUCachePolicy.h#L18), we don't directly know how many marks are inside each MarksInCompressedFile without accessing its [num_marks field](https://github.com/ClickHouse/ClickHouse/blob/759349e6a2fb0f15506c5ff9abc71ac149042db1/src/Formats/MarkInCompressedFile.h#L99).

To compute `total_evicted_marks_num`, it seems we would need to access `num_marks`, which might require including `MarksInCompressedFile` in `SLRUCachePolicy.h`. That could break the abstraction and increase coupling.

Would you have an alternative suggestion for how to compute `total_evicted_marks_num` cleanly without breaking this separation?

shiv4289 committed:
Update to match style guide as suggested in review

 Co-authored-by: Kseniia Sumarokova <54203879+kssenii@users.noreply.github.com>

shiv4289 committed:
Make eviction stats lighter using on_remove() callback

Earlier we were keeping complete evicted value, now we only keep the size_t counters

shiv4289 commented:
[Merged with main](https://github.com/ClickHouse/ClickHouse/compare/4bccea62a8a4366405fde19df813426d557a47a8..c387ef182097c61665f5105533a29d873bf04429) to remove conflicts that were showing up in this PR after https://github.com/ClickHouse/ClickHouse/pull/81023 was merged to main using
```
git pull origin master --rebase
git push --force
```
History of [commits in this PR](https://github.com/ClickHouse/ClickHouse/pull/80799/commits) are identical so as to not disturb the review flow. Will keep an eye on the CI.

shiv4289 committed:
remove redundant onRemoveOverflowWeightLoss()

shiv4289 committed:
improve variable names

Removed mark cache references from variable names in Cachebase.h, make it more generic

shiv4289 committed:
removed onEviction() method and EvictionStats struct

shiv4289 commented on src/Storages/MergeTree/VectorSimilarityIndexCache.h:
```diff
@@ -89,9 +89,10 @@ class VectorSimilarityIndexCache : public CacheBase<UInt128, VectorSimilarityInd
 private:
-    void onRemoveOverflowWeightLoss(size_t weight_loss) override
+    /// Called when cache eviction occurs
+    void onEviction(const EvictionStats& stats) override
```
This line is dropped now.

shiv4289 commented on src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadataFilesCache.h:
```diff
@@ -150,9 +150,10 @@ class IcebergMetadataFilesCache : public CacheBase<String, IcebergMetadataFilesC
 private:
-    void onRemoveOverflowWeightLoss(size_t weight_loss) override
+    /// Called when cache eviction occurs
+    void onEviction(const EvictionStats& stats) override
```
This line is dropped now.

kssenii commented on src/Common/PageCache.cpp:
```diff
@@ -145,9 +145,10 @@ bool PageCache::contains(const PageCacheKey & key, bool inject_eviction) const
 
-void PageCache::Shard::onRemoveOverflowWeightLoss(size_t weight_loss)
+void PageCache::Shard::onValueRemoval(const MappedPtr & mappedPtr)
 {
-    ProfileEvents::increment(ProfileEvents::PageCacheWeightLost, weight_loss);
+    ProfileEvents::increment(ProfileEvents::PageCacheWeightLost);
```
Need to pass `cell.size` here along with `cell.value` and increment profile event by this size, otherwise this place is incorrect (increment by 1 is not what we want here).

shiv4289 replied:
Please check [comment below](https://github.com/ClickHouse/ClickHouse/pull/80799#discussion_r2135914535) for a clarification on this.

shiv4289 commented on src/Storages/MergeTree/VectorSimilarityIndexCache.h:
```diff
@@ -89,9 +89,11 @@ class VectorSimilarityIndexCache : public CacheBase<UInt128, VectorSimilarityInd
+    void onValueRemoval(const MappedPtr & mappedPtr) override
     {
-        ProfileEvents::increment(ProfileEvents::VectorSimilarityIndexCacheWeightLost, weight_loss);
+        ProfileEvents::increment(ProfileEvents::VectorSimilarityIndexCacheWeightLost);
```
Yes, that’s an oversight on my part—thanks for catching it (and sorry about that). The intent was to use the weight function for that cache. I’ll fix it quickly.

shiv4289 commented on src/Storages/MarkCache.h:
```diff
@@ -60,21 +60,17 @@ class MarkCache : public CacheBase<UInt128, MarksInCompressedFile, UInt128Trivia
+
         auto marks_in_compressed_file = std::static_pointer_cast<MarksInCompressedFile>(mappedPtr);
-        return marks_in_compressed_file->getNumberOfMarks();
+        ProfileEvents::increment(ProfileEvents::MarkCacheEvictedBytes, MarksWeightFunction()(*marks_in_compressed_file));
```
@kssenii @rschu1ze  does this sound okay to you? This is what I originally intended, which is why I didn’t pass cell.size explicitly. I’ll also double-check whether every cache type has a dedicated and easily accessible weight function—we should make sure that’s consistent.

shiv4289 replied:
discussing this [here](https://github.com/ClickHouse/ClickHouse/pull/80799#discussion_r2137259135)

shiv4289 committed:
style changes

shiv4289 commented on src/Common/PageCache.cpp:
```diff
@@ -145,9 +145,10 @@ bool PageCache::contains(const PageCacheKey & key, bool inject_eviction) const
+void PageCache::Shard::onValueRemoval(const MappedPtr & mapped_ptr)
 {
-    ProfileEvents::increment(ProfileEvents::PageCacheWeightLost, weight_loss);
+    auto page_cache_cell = std::static_pointer_cast<PageCacheCell>(mapped_ptr);
```
Hi @rschu1ze , the other suggestion (from @kssenii [here](https://github.com/ClickHouse/ClickHouse/pull/80799#discussion_r2142164330) ) is to use the result (weight lost) passed from CachePolicy's removeOverflow() method back to the cachebase in the onEntryRemoval() (previously OnValueRemoval()) method signature along with mapped_ptr (cache value). 
Thoughts? If you agree with that suggestion, I can go with it and the current conversation can be resolved.

shiv4289 replied:
This line of code is removed as discussed.

shiv4289 committed:
readability improvements as per review feedback

shiv4289 committed:
prefer passing calculated weight loss param against repeat method calls

shiv4289 commented on src/Common/CacheBase.h:
```diff
@@ -64,7 +64,10 @@ class CacheBase
-        auto on_weight_loss_function = [&](size_t weight_loss) { onRemoveOverflowWeightLoss(weight_loss); };
+        auto on_remove_function = [this](const MappedPtr & mapped_ptr)
+        {
+            onValueRemoval(mapped_ptr);
```
Changed as suggested.

shiv4289 commented on src/Common/CacheBase.h:
```diff
@@ -335,8 +338,9 @@ class CacheBase
 
-    /// Override this method if you want to track how much weight was lost in removeOverflow method.
-    virtual void onRemoveOverflowWeightLoss(size_t /*weight_loss*/) {}
+    /// This is called when a cell is being evicted from the cache.
```
Changed as suggested.

kssenii commented on src/Common/SLRUCachePolicy.h:
```diff
@@ -326,16 +326,16 @@ class SLRUCachePolicy : public ICachePolicy<Key, Mapped, HashFunction, WeightFun
             else
             {
                 current_weight_lost += cell.size;
+                if (!is_protected)
```
We already checked that `entry.is_protected == false` in this scope, this means that this arg `is_protected` will equal false as well always, as we cannot have protected cells in non-protected queue and the other way round. I'd remove it completely or add `chassert(!is_protected)`.

shiv4289 committed:
fix the event update condition in SLRU cache

shiv4289 commented:
Merged master to the PR branch.

shiv4289 committed:
Use static cast against static_pointer_cast

Co-authored-by: Kseniia Sumarokova <54203879+kssenii@users.noreply.github.com>
